File: 0100-same-tree/0100-same-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function(p, q) {
  if (!p && !q) {
    return true;
  }

  if (!p || !q) {
    return false;
  }
  
  if (p.val !== q.val) {
    return false;
  }

  if ((!p.left && q.left) || (p.left && !q.left)) {
    return false;
  }
  
  if (p.left && q.left) {
    if (!isSameTree(p.left, q.left)) {
      return false; 
    };
  }


  if ((!p.right && q.right) || (p.right && !q.right)) {
    return false;
  }

  if (p.right && q.right) {
    if (!isSameTree(p.right, q.right)) {
      return false; 
    };
  }

  return true;
};

File: 0100-same-tree/0100-same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p and q:
            if p.val != q.val:
                return False
        elif p or q:
            return False
        else:
            return True

        if p.left and q.left:
            if self.isSameTree(p.left, q.left) == False:
                return False
        elif p.left or q.left:
            return False

        if p.right and q.right:
            if self.isSameTree(p.right, q.right) == False:
                return False
        elif p.right or q.right:
            return False
        return True